### PR TITLE
Accepted all-item-types option, added ACL control, updated interface

### PR DESCRIPTION
**Pull request by DBinaghi**
Changes summary (rules logic has not been touched):
- adapted style to Omeka Classic version 3.0
- added explanations for adding/editing rules
- added "All types" option, as a rule could apply to all Items
- added confirm messages for adding/editing actions
- improved rule description on deleting rule
- added ACL control to limit access to the plugin configuration
- sorted element list
- fixed (https://github.com/biblibre/omeka-plugin-ItemDuplicateCheck/issues/8) issue
- updated readme and languages files (general, Italian and French)

In order to test confirm messages texts: create a dummy rule, then edit dummy rule, then delete dummy rule.
In order to test ACL control: login as user with level lower than admin.

### CODE REVIEW

**Comment by jajm** [comment]
Thanks for adding changes summary in PR description, that will help a lot!
Let me know when it is ready for review

File: ItemDuplicateCheckPlugin.php [hunk 1]
```diff
@@ -6,11 +6,12 @@ class ItemDuplicateCheckPlugin extends Omeka_Plugin_AbstractPlugin
...
         'uninstall',
         'initialize',
-        'admin_head',
+        'define_acl',
+        'admin_head'
     );
 
     protected $_filters = array(
-        'admin_navigation_main',
+        'admin_navigation_main'
```
Comment (jajm) [inline ItemDuplicateCheckPlugin.php]: As a general rule, please do not modify what doesn't need to be modified.
Removing the comma here is useless, and it adds noise to the PR

File: views/admin/check/check.php [hunk 2]
```diff
@@ -1,22 +1,19 @@
...
-                echo __('Duplicate found');
-            } else {
-                echo __('Duplicates found');
-            }
-        ?>:
-        <ul>
+            $countDuplicates = count($duplicates);
+            echo __(plural("Warning: a duplicate of this Item has been found!", "Warning: %s duplicates of this Item have been found!", $countDuplicates), $countDuplicates);
+        ?>
+        <ul style="list-style-type: none; margin-left: 0; padding-left: 0";>
```
Comment (jajm) [inline views/admin/check/check.php]: There is a `;` outside of the style attribute

File: views/admin/check/check.php [hunk 3]
```diff
@@ -1,22 +1,19 @@
...
             <?php foreach ($duplicates as $duplicate): ?>
                 <?php $dItem = $duplicate['item']; ?>
                 <?php $dRule = $duplicate['rule']; ?>
                 <li>
-                    <a href="<?php echo url('items') . '/show/' . $dItem->id; ?>"><?php echo metadata($dItem, array('Dublin Core', 'Title')); ?> (#<?php echo $dItem->id; ?>)</a>
-                    <ul>
+                    #<?php echo $dItem->id; ?>: <a href="<?php echo url('items') . '/show/' . $dItem->id; ?>"><?php echo metadata($dItem, array('Dublin Core', 'Title')); ?></a>
+                    <ul style="list-style-type: none";>
```
Comment (jajm) [inline views/admin/check/check.php]: There is a `;` outside of the style attribute here too

File: controllers/RulesController.php [hunk 4]
```diff
@@ -6,96 +6,104 @@ public function listAction()
...
+    
     protected function _getElements()
     {
+        $show_element_set_headings = get_option('show_element_set_headings');
         $db = get_db();
         $table = $db->getTable('Element');
         $options = $table->findPairsForSelectForm(array(
             'record_types' => array('Item', 'All'),
-            'sort' => 'orderBySet'
+            'sort' => ($show_element_set_headings ? 'alphaBySet' : 'alpha')
```
Comment (jajm) [inline controllers/RulesController.php]: I think it's better to use `orderBySet` here, to let users be able to configure the order they want.

File: controllers/RulesController.php [hunk 5]
```diff
@@ -6,96 +6,104 @@ public function listAction()
...
+            foreach ($options as $option) {
+                $optGroup = $option['item_type_name']
+                    ? __('Item Type') . ': ' . __($option['item_type_name'])
+                    : __($option['element_set_name']);
+                $value = __($option['element_name']);
+                if ($value != '') $options[$optGroup][$option['element_id']] = $value;
+            }
+            // sort alphabetically element names in each element set
+            foreach ($options as &$option) {
+                asort($option);
```
Comment (jajm) [inline controllers/RulesController.php]: I don't know what this is supposed to do, but for me it only emits PHP warnings. If I remove the block I don't see any change (except that there are no warnings anymore)

File: languages/fr.po [hunk 6]
```diff
@@ -1,63 +1,127 @@
...
+
+#: ItemDuplicateCheckPlugin.php:52
+msgid "Item Duplicate Check"
+msgstr "Contrôle de contenu doublon"
+
+#: views/admin/check/check.php:5
+msgid "Warning: a duplicate of this Item has been found!"
+msgid_plural "Warning: %s duplicates of this Item have been found!"
+msgstr[0] "Attention : un doublon de cet contenu a été trouvé !"
+msgstr[1] "Attention : %s doublons de cet contenu ont été trouvés !"
```
Comment (jajm) [inline languages/fr.po]: It should be 'ce contenu', not 'cet contenu'

**Review by jajm — CHANGES_REQUESTED** [review]
Overall, changes are looking good!
There are only a few things to fix. If you can do that, I'll merge it asap

File: ItemDuplicateCheckPlugin.php (hunk 1 above)
Comment (DBinaghi) [reply ItemDuplicateCheckPlugin.php]: No problem for me, of course, but I do think the comma after the last element is always useless, and is just ignored by the interpreter, so I normally don't add it.

File: views/admin/check/check.php (hunk 2 above)
Comment (DBinaghi) [reply views/admin/check/check.php]: Thx

File: views/admin/check/check.php (hunk 3 above)
Comment (DBinaghi) [reply views/admin/check/check.php]: Thx

File: controllers/RulesController.php (hunk 4 above)
Comment (DBinaghi) [reply controllers/RulesController.php]: I am actually working on that: if "show element sets names" is not on, there's just a list of element names, I think it's clearer for user to let the user know which element set they belong to. I'll submit my idea asap.

File: controllers/RulesController.php (hunk 5 above)
Comment (DBinaghi) [reply controllers/RulesController.php]: Once elements groups are created, then sorting them alphabetically solves a problem: without that, using translated element names messes up with the order. I did not notice any warning in my version, could you copy/paste any of them to let me understand what's causing them?

File: languages/fr.po (hunk 6 above)
Comment (DBinaghi) [reply languages/fr.po]: Thx (Google translator to blame, this time :))

File: controllers/RulesController.php (hunk 5 above)
Comment (jajm) [reply controllers/RulesController.php]: This is what I see on `/admin/item-duplicate-check/rules/add`:

```
Notice: Undefined index: item_type_name in /home/omeka/omeka/plugins/ItemDuplicateCheck/controllers/RulesController.php on line 92
Notice: Undefined index: element_set_name in /home/omeka/omeka/plugins/ItemDuplicateCheck/controllers/RulesController.php on line 94
Notice: Undefined index: element_name in /home/omeka/omeka/plugins/ItemDuplicateCheck/controllers/RulesController.php on line 95
Notice: Undefined index: item_type_name in /home/omeka/omeka/plugins/ItemDuplicateCheck/controllers/RulesController.php on line 92
Notice: Undefined index: element_set_name in /home/omeka/omeka/plugins/ItemDuplicateCheck/controllers/RulesController.php on line 94
Notice: Undefined index: element_name in /home/omeka/omeka/plugins/ItemDuplicateCheck/controllers/RulesController.php on line 95
Notice: Undefined index: item_type_name in /home/omeka/omeka/plugins/ItemDuplicateCheck/controllers/RulesController.php on line 92
Notice: Undefined index: element_set_name in /home/omeka/omeka/plugins/ItemDuplicateCheck/controllers/RulesController.php on line 94
Notice: Undefined index: element_name in /home/omeka/omeka/plugins/ItemDuplicateCheck/controllers/RulesController.php on line 95
Notice: Undefined index: item_type_name in /home/omeka/omeka/plugins/ItemDuplicateCheck/controllers/RulesController.php on line 92
Notice: Undefined index: element_set_name in /home/omeka/omeka/plugins/ItemDuplicateCheck/controllers/RulesController.php on line 94
Notice: Undefined index: element_name in /home/omeka/omeka/plugins/ItemDuplicateCheck/controllers/RulesController.php on line 95
```

Comment (jajm) [reply controllers/RulesController.php]: Note that the comment I made is not only about the sort. It is also about the block above that is not displayed here :)

File: controllers/RulesController.php (hunk 4 above)
Comment (DBinaghi) [reply controllers/RulesController.php]: After thinking about your comment, but also taking into consideration the possibility of admin wanting a different behaviour for this plugin as opposed to the site's default behaviour, I think the best option would be to introduce a config option (incidentally: there's no config page now, but I've created it as there's at least another two options that could be added) that allows to choose which Element List layout to use. What do you reckon?

Comment (jajm) [reply controllers/RulesController.php]: That would be great. But if you want to work on it, please do it in another pull request (create a new branch first). There are already several unrelated changes here. Let's merge these ones before adding something else ;)

Comment (DBinaghi) [reply controllers/RulesController.php]: Rolled back to original function, for this branch. It should be good for mergin, now. In the meantime, I'll be working on the other changes.

File: controllers/RulesController.php (hunk 5 above)
Comment (DBinaghi) [reply controllers/RulesController.php]: Rolled back to original function, for this branch

**Pull request merged**